### PR TITLE
feat: Support installing multi-language environment at the same time

### PR DESCRIPTION
**Pull request by oubotong**
This is required by R because if you want to use tensorflow with R, it will require python environment.
Also, it is better to allow users to install multiple languages environment at the same time.
Related to #407, #1344

### CODE REVIEW

File: pkg/lang/ir/v1/system.go [hunk 1]
```diff
@@ -231,40 +231,44 @@ func (g *generalGraph) compileExtraSource(root llb.State) (llb.State, error) {
...
+				pack = llb.Merge([]llb.State{
+					root,
+					llb.Diff(root, pypi, llb.WithCustomName("[internal] PyPI packages")),
+					llb.Diff(root, conda, llb.WithCustomName("[internal] conda packages")),
+				}, llb.WithCustomName("[internal] Python packages"))
+			}
+		case "r":
+			pack = g.installRPackages(pack)
+		case "julia":
+			pack = g.installJuliaPackages(pack)
```
Comment (kemingy) [inline pkg/lang/ir/v1/system.go:270]: Can they run in parallel?

**Review by terrytangyuan — COMMENTED** [review]
Related issues https://github.com/tensorchord/envd/issues/407 https://github.com/tensorchord/envd/issues/1344

File: pkg/lang/ir/v1/python.go [hunk 2]
```diff
@@ -68,11 +68,17 @@ func (g *generalGraph) installPython(root llb.State) (llb.State, error) {
...
 
 func (g generalGraph) getAppropriatePythonVersion() (string, error) {
-	if g.Language.Version == nil {
-		return PythonVersionDefault, nil
+	var version string
+	for _, language := range g.Languages {
+		if language.Name == "python" {
+			if language.Version == nil {
+				return PythonVersionDefault, nil
+				version = *language.Version
```
[the remark this answers is not in the document]
Comment (oubotong) [reply pkg/lang/ir/v1/python.go]: Fixed

File: pkg/lang/ir/v1/system.go [hunk 3]
```diff
@@ -231,40 +231,44 @@ func (g *generalGraph) compileExtraSource(root llb.State) (llb.State, error) {
...
+	for _, language := range g.Languages {
+		switch language.Name {
+		case "python":
+			lang, err = g.installPython(lang)
+		case "r":
+			rSrc := g.compileRLang(lang)
+			lang = g.installRLang(rSrc)
+		case "julia":
+			lang = g.installJulia(lang)
+		}
```
[the remark this answers is not in the document]
Comment (oubotong) [reply pkg/lang/ir/v1/system.go]: Done

File: pkg/lang/ir/v1/system.go [hunk 4]
```diff
@@ -313,10 +317,12 @@ func (g *generalGraph) compileBaseImage() (llb.State, error) {
...
 		"image":    g.Image,
-		"language": g.Language.Name,
+		"language": g.Languages,
 	})
-	if g.Language.Version != nil {
-		logger = logger.WithField("version", *g.Language.Version)
+	for _, language := range g.Languages {
+		if language.Version != nil {
+			logger = logger.WithField("version", language.Version)
+		}
```
[the remark this answers is not in the document]
Comment (oubotong) [reply pkg/lang/ir/v1/system.go]: Removed

File: pkg/lang/ir/v1/system.go [hunk 5]
```diff
@@ -229,44 +230,63 @@ func (g *generalGraph) compileExtraSource(root llb.State) (llb.State, error) {
 }
 
 func (g *generalGraph) compileLanguage(root llb.State) (llb.State, error) {
+	var wg sync.WaitGroup
```
Comment (kemingy) [inline pkg/lang/ir/v1/system.go:233]: You don't need to use the goroutine and wait group here. The BuildKit LLB only builds the DAG graph and executes it later.

File: pkg/lang/ir/v1/system.go [hunk 6]
```diff
@@ -229,44 +230,63 @@ func (g *generalGraph) compileExtraSource(root llb.State) (llb.State, error) {
...
+		}(language)
 	}
-
-	return lang, err
+	wg.Wait()
+	return llb.Merge(langs, llb.WithCustomName("[internal] build all language environments")), err
 }
 
 func (g *generalGraph) compileLanguagePackages(root llb.State) llb.State {
+	var wg sync.WaitGroup
```
Comment (kemingy) [inline pkg/lang/ir/v1/system.go:258]: The same here.

File: pkg/lang/ir/v1/system.go (hunk 5 above)
Comment (oubotong) [reply pkg/lang/ir/v1/system.go]: Updated

File: pkg/lang/ir/v1/system.go (hunk 6 above)
Comment (oubotong) [reply pkg/lang/ir/v1/system.go]: Updated

**Review by kemingy — APPROVED** [review]

**Pull request merged**